### DISCUSSION TITLE
Issues building a cassette tape model

### DISCUSSION THREAD

**Discussion by pokepress** [pokepress/aero#35] [General]
So, over the last few months, I've been working on developing a cassette-tape model. I've acquired a three-head deck that supports type I, II, and IV tapes, and has Dolby B, C, and S noise reduction. The issue I'm having is getting the model to remove noise without removing detail, particularly on the high-end. This issue is at its worst when using Type I (ferric) tapes with no NR (common for most releases today), as the noise floor is not particularly good.

I've attempted to compensate using the stft_mag_weight_shift setting, which is designed to allow skewing the model toward reducing sounds that are too loud (negative values) or increasing the volume of sounds that are too low (positive numbers). This effectively alters the numeric tradeoff between noise reduction and frequency retention/reconstruction. For cassettes, I've set this to an absurdly high value (>= .99), which does help more sound get kept, but also causes the model to skew high on volumes (the processed audio has noticeably excessive bass), and doesn't result in great high-end fidelity, particularly for music.

One way to mitigate this would be to create a distinct "music" dataset, but overall I know I'm really just pushing weight shift parameter way too far and should make the STFT loss (or perhaps a new loss) more intelligent to avoid these issues, because right now you're better off just using the noise reduction filter in your audio editor to remove tape hiss.

Any suggestions?

**Comment by pokepress** [comment]
After looking into the code, I think part of the issue is this line in the STFT loss code:

`conv_loss = torch.mean(pow(scaled_diff,1.3)*torch.maximum(x_mag,y_mag))`

In general, it makes sense to weight the loss by the higher of the two magnitudes, however this can cause a scenario where the model spends too much effort on reducing noise at the expense of detail, particularly in the higher frequencies. I've been trying this:

`conv_loss = torch.mean(pow(scaled_diff,1.3)*torch.maximum(x_mag/4,y_mag))`

So that the model gets less "credit" for cleaning up low-level noise. It seems to help keep higher frequencies intact, but the numerical relationship should probably be be something more precise than reducing the inferred magnitude.
